src/int.cr

```diff
@@ -357,8 +357,47 @@ struct Int
     (self & mask) == mask
   end
 
-  def gcd(other : Int)
-    self == 0 ? other.abs : (other % self).gcd(self)
+  # Returns the greatest common divisor of `self` and `other` (which
+  # is an `Int` of some kind). The return type is whichever side can
+  # fit the biggest values. If either is signed and has value equal to
+  # `MIN` of its type, then overflow will be raised.
+  #
+  # ```
+  # 5.gcd(10) # => 2
+  # 5.gcd(7)  # => 1
+  # ```
+  def gcd(other : self) : self
+    u = self.abs
+    v = other.abs
+    shift = self.class.zero
+    return v if u == 0
+    return u if v == 0
+
+    # Let shift := lg K, where K is the greatest power of 2
+    # dividing both u and v.
+    while (u | v) & 1 == 0
```

Comment:
Oddly enough, using `odd?`  actually gave noticeably worse performance. I tried to replace the implementation of `odd?` to be this under the hood, but benchmarking that did not give any noticeable difference that I could see. :man_shrugging:

Reply:
Just for interest - try adding `AlwaysInline` to the `odd?` implementation. This might just be a LLVM inlining heuristics filure.

Reply:
Yeah, no difference. Replacing the method helps `gcd` for everything but `UInt8 ` (and by help, I mean it makes the whole difference if it is categorically faster than the old one or several times slower), but I'm not managing to replicate that performance difference benchmarking `odd?` itself. There is some improvement for signed numbers, but not for unsigned, while the performance difference shows in both. 

All regardless of `@[AlwaysInline]` or not.

Reply:
This loop is doing `shift = {u.trailing_zeroes, v.trailing_zeroes}.min` right? Perhaps we could bind the `llvm.cttz` intrinsics in the future for more speed :)

Reply:
We already have `cttz` bound at `Int#trailing_zeros_count`. However, I tried it and didn't notice any noticable improvement, so I left that out. Which isn't all that surprising, as on average, most numbers don't really have lots of trailing zeros.

But perhaps I did something wrong in my implementation, or it could be affected by me having a quite old desktop computer which I used to test it.. :man_shrugging: